packages/rpc-transport/src/transports/websocket/websocket-connection.ts

```diff
@@ -0,0 +1,135 @@
+                    new Error('WebSocket failed to connect', { cause: ev })
+                );
+            }
+        }
+        let hasConnected = false;
+        function handleOpen() {
+            hasConnected = true;
+            resolve({
+                async send(payload: unknown) {
+                    const message = JSON.stringify(payload);
+                    webSocket.send(message);
+                },
+                async *[Symbol.asyncIterator]() {
+                    const iteratorKey = Symbol();
+                    iteratorState.set(iteratorKey, { __hasPolled: false, queuedMessages: [] });
+                    try {
+                        while (true) {
+                            const state = iteratorState.get(iteratorKey);
+                            if (!state) {
+                                // There should always be state by now.
+                                throw new Error('Invariant: WebSocket message iterator is missing state storage');
+                            }
+                            if (state.__hasPolled) {
+                                // You should never be able to poll twice in a row.
+                                throw new Error(
+                                    'Invariant: WebSocket message iterator state is corrupt; ' +
+                                        'iterated without first resolving existing message promise'
+                                );
+                            }
+                            const queuedMessages = state.queuedMessages;
+                            if (queuedMessages.length) {
+                                state.queuedMessages = [];
+                                yield* queuedMessages;
+                            } else {
+                                try {
+                                    yield await new Promise((onMessage, onError) => {
+                                        iteratorState.set(iteratorKey, {
+                                            __hasPolled: true,
+                                            onError,
+                                            onMessage,
+                                        });
+                                    });
+                                } catch (e) {
+                                    if (
+                                        e !== null &&
+                                        typeof e === 'object' &&
+                                        'type' in e &&
+                                        e.type === 'close' &&
+                                        'wasClean' in e &&
+                                        e.wasClean
+                                    ) {
+                                        return;
+                                    } else {
+                                        // TODO: Coded error.
+                                        throw new Error('WebSocket connection closed', { cause: e });
```

Comment:
I went digging on coded errors and am not sure if this is where its relevant, but I found this snippet:

https://websockets.spec.whatwg.org/#dom-websocket-onerror

Reply:
Oh, by coded error I mean we need to do [this](https://github.com/solana-labs/solana-web3.js/blob/630f6d22ab7e779092a71c390cb5b7fe5f198253/packages/library-legacy/src/errors.ts#L13-L50), but for every single error in this repository.

That way, when someone _receives_ an error thrown, they don't have to do this:

```ts
// You had to _import_ the concrete implementation of `SolanaReallySpecificError`
// just to do this comparison - even if your app _never_ fails.
if (e instanceof SolanaReallySpecificError) { /*  ... * }
```

They can just do this:

```ts
if (
  e instanceof Error &&
  e.name === 'SolanaClientError' &&
  e.code === SolanaClientErrorCode.REALLY_SPECIFIC_ERROR
) { /* ... */ }
```

Typescript will inline the error code number and throw away the import, and if it doesn't you can always do this:

```ts
if (
  e instanceof Error &&
  e.name === 'SolanaClientError' &&
  e.code === 42069 satisfies (typeof SolanaClientErrorCode)['REALLY_SPECIFIC_ERROR']
) { /* ... */ }
```

Reply:
tehe look what this made me find: https://github.com/solana-labs/solana/commit/a2cf25153fe1bcd5b81a776f89bb34de0cef7547